### PR TITLE
[Messenger] use database platform to convert correctly the DateTime

### PR DESCRIPTION
**Pull request by roukmoute**
| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes 
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/issues/32427
| License       | MIT

In Doctrine Messenger the method `\Symfony\Component\Messenger\Transport\Doctrine\Connection::formatDateTime()` is used to format dateTime into this: `Y-m-d\TH:i:s`.
But this is not supported in all databases platform.

Here we use the database platform to convert correctly the dateTime.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 1]
```diff
@@ -122,8 +123,8 @@ public function send(string $body, array $headers, int $delay = 0): string
             ':body' => $body,
             ':headers' => json_encode($headers),
             ':queue_name' => $this->configuration['queue_name'],
-            ':created_at' => self::formatDateTime($now),
-            ':available_at' => self::formatDateTime($availableAt),
+            ':created_at' => $this->convertDateTime($now),
+            ':available_at' => $this->convertDateTime($availableAt),
```
Comment (stof) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: We could even use the datetime directly as parameter and pass the Doctrine DBAL types (in the third argument) for these, so that DBAL handles the conversion calls for us (and would avoid the need to create a public `convertDateTime` too, as tests can do the same)

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 2]
```diff
@@ -329,9 +330,21 @@ private function getSchema(): Schema
         return $schema;
     }
 
-    public static function formatDateTime(\DateTimeInterface $dateTime)
+    /**
+     * @throws DBALException
+     *
+     * @deprecated since Symfony 4.4, use convertDateTime instead
+     */
+    public static function formatDateTime(\DateTimeInterface $dateTime, AbstractPlatform $platform)
```
Comment (stof) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: adding a required argument is a BC break, which is a bad idea for a method kept as a BC layer...

Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: Oh yes, I will add a new commit to fix this.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 3]
```diff
@@ -329,9 +330,25 @@ private function getSchema(): Schema
         return $schema;
     }
 
-    public static function formatDateTime(\DateTimeInterface $dateTime)
+    /**
+     * @throws DBALException
+     *
+     * @deprecated since Symfony 4.4, use convertDateTime instead
+     */
+    public static function formatDateTime(\DateTimeInterface $dateTime, AbstractPlatform $platform = null)
```
Comment (derrabus) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: You've added a second parameter to the method, but I can find any calls with that parameter being set.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 3 above)
Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: See my commits in order and you will understand my logic :)

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 1 above)
Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: Sorry I don't understand.  
Because in this file, the `executeQuery` method is not the same method of https://github.com/doctrine/dbal/blob/66f28111d57fdc3508d4c97989257291702176cf/lib/Doctrine/DBAL/Connection.php#L883.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 4]
```diff
@@ -329,9 +330,28 @@ private function getSchema(): Schema
         return $schema;
     }
 
-    public static function formatDateTime(\DateTimeInterface $dateTime)
+    /**
+     * @throws DBALException
+     *
+     * @deprecated since Symfony 4.4, use convertDateTime instead
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: the deprecation must be done in a separate PR, targetting 4.4 (this PR is for 4.3)

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: that's a BC break, can't be done

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 1 above)
Comment (stof) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: it is the same. The `driverConnection` propert contains a `Doctrine\DBAL\Connection`, which is exactly the code I linked.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 3 above)
Comment (stof) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: it is not a question of commits in order. The existing method that you want to deprecate must not change signature.

Comment (stof) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: you should remove the new argument and always use `$dateTime->format('Y-m-d\TH:i:s')` in this deprecated method. Your current change still requires changing the caller to add the platform to benefit from the fix, at which point it is as easy to switch to the new method instead.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 5]
```diff
@@ -329,9 +330,28 @@ private function getSchema(): Schema
...
+
+        if (null === $platform) {
+            return $dateTime->format('Y-m-d\TH:i:s');
+        }
+
+        return Type::getType(Type::DATETIME)->convertToDatabaseValue($dateTime, $platform);
+    }
+
+    /**
+     * @throws DBALException
```
Comment (stof) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: should be removed. We don't care about this one here (and anyway, this API should not be public, so this would be useless)

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 1 above)
Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: I think @stof means we could use `\Doctrine\DBAL\Connection::executeQuery` which has the types as parameter.

**Review by Tobion — CHANGES_REQUESTED** [review]
https://github.com/symfony/symfony/pull/32456#discussion_r301505428
@roukmoute do you have time to finish the PR?

**Comment by roukmoute** [comment]
yep, I've finally received my laptop so I can fix it this week-end

File: src/Symfony/Component/Messenger/composer.json [hunk 6]
```diff
@@ -20,6 +20,7 @@
         "psr/log": "~1.0"
     },
     "require-dev": {
+        "ext-json": "*",
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/composer.json]: what does messenger require json for? also this is an unrelated changed.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 7]
```diff
@@ -118,12 +119,15 @@ public function send(string $body, array $headers, int $delay = 0): string
                 'available_at' => ':available_at',
             ]);
 
-        $this->executeQuery($queryBuilder->getSQL(), [
+        $this->driverConnection->executeQuery($queryBuilder->getSQL(), [
             ':body' => $body,
             ':headers' => json_encode($headers),
             ':queue_name' => $this->configuration['queue_name'],
-            ':created_at' => self::formatDateTime($now),
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: please remove the `formatDateTime` method

Comment (derrabus) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: Can we do this in a bugfix release? After all, the method is declared public. 😕

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 8]
```diff
@@ -118,12 +119,15 @@ public function send(string $body, array $headers, int $delay = 0): string
                 'available_at' => ':available_at',
             ]);
 
-        $this->executeQuery($queryBuilder->getSQL(), [
+        $this->driverConnection->executeQuery($queryBuilder->getSQL(), [
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: you lost the logic of the local `executeQuery` everywhere (catching `TableNotFoundException`) which we still need

File: src/Symfony/Component/Messenger/composer.json (hunk 6 above)
Comment (roukmoute) [reply src/Symfony/Component/Messenger/composer.json]: Because it uses `json_decode` and `json_encode`.
Yeah it is not related with this P.R., but it is so minimalist, I supposed it could pass.

Would you like I remove this commit?

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 7 above)
Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: I was thinking I delete it into another P.R. for the 4.4 branch to do not do a BC Break?

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 8 above)
Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: This is related to: https://github.com/symfony/symfony/pull/32456#discussion_r301505428

All tests passing, so a test is currently missing, right?

What would you do except?

File: src/Symfony/Component/Messenger/composer.json (hunk 6 above)
Comment (Tobion) [reply src/Symfony/Component/Messenger/composer.json]: Yes. json is only used in certain cases like redis. So it's an optional dependency.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 7 above)
Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: can be removed according to https://github.com/symfony/symfony/pull/32640#discussion_r305633648

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 8 above)
Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: you can use `$this->driverConnection->executeQuery` within `$this->executeQuery`

**Comment by roukmoute** [comment]
@Tobion Do you have any idea about the failure tests?

**Comment by roukmoute** [comment]
@Tobion Any news?

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 9]
```diff
@@ -329,15 +333,19 @@ private function getSchema(): Schema
...
     private function decodeEnvelopeHeaders(array $doctrineEnvelope): array
     {
         $doctrineEnvelope['headers'] = json_decode($doctrineEnvelope['headers'], true);
 
         return $doctrineEnvelope;
     }
+
+    private function getDateTimeType()
+    {
+        if (!class_exists('\Doctrine\DBAL\Types\Types')) {
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: please use `\Doctrine\DBAL\Types\Types::class`

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 10]
```diff
@@ -329,15 +333,19 @@ private function getSchema(): Schema
...
+            return Type::DATETIME;
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: this one seems to be deprecated. so it's good that you used the new constant. but `getSchema` still uses hard-codes the deprecated constant. so it needs to be replaced there as well.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 11]
```diff
@@ -268,9 +274,12 @@ private function createAvailableMessagesQueryBuilder(): QueryBuilder
             ->andWhere('m.available_at <= :now')
             ->andWhere('m.queue_name = :queue_name')
             ->setParameters([
-                ':now' => self::formatDateTime($now),
+                ':now' => $now,
                 ':queue_name' => $this->configuration['queue_name'],
-                ':redeliver_limit' => self::formatDateTime($redeliverLimit),
+                ':redeliver_limit' => $redeliverLimit,
+            ], [
+                ':now' => Type::getType($this->getDateTimeType()),
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: do you really need to pass the type instance here with `Type::getType`? I don't think so. Passing the constant value should be enough.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 12]
```diff
@@ -281,22 +290,17 @@ private function createQueryBuilder(): QueryBuilder
             ->from($this->configuration['table_name'], 'm');
     }
 
-    private function executeQuery(string $sql, array $parameters = [])
+    private function executeQuery(string $sql, array $parameters = [], ?array $types = [])
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: why does $types need to be nullable?

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 10 above)
Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: If I do not use this method, tests does not passes.
Don't care?

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 12 above)
Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: Without that, few tests does not passes.

```sh
1) Symfony\Component\Messenger\Tests\Transport\Doctrine\ConnectionTest::testGetAMessageWillChangeItsStatus
TypeError: Argument 3 passed to Symfony\Component\Messenger\Transport\Doctrine\Connection::executeQuery() must be of the type array, null given, called in /home/roukmoute/workspace/symfony/src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php on line 150
```

Would you prefer I change theses tests to pass without this nullable?

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 10 above)
Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: please fix the tests then

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 12 above)
Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: The mocks probably need to be fixed so it's does not return a null which violates the contract.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 10 above)
Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: I can't it comes from versions used by travis.
I will remove this and you will see :)

Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: You must have misunderstood me. `getDateTimeType` using the fallback was fine. But the method was not used consistently everywhere. I don't see why you removed this now.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 13]
```diff
@@ -306,20 +314,20 @@ private function getSchema(): Schema
     {
         $schema = new Schema([], [], $this->driverConnection->getSchemaManager()->createSchemaConfig());
         $table = $schema->createTable($this->configuration['table_name']);
-        $table->addColumn('id', Type::BIGINT)
+        $table->addColumn('id', Types::BIGINT)
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: you cannot require `Types` as it's not released in dbal yet. please just keep using `Type` everywhere. we can handle the deprecation of it in the future when it's actually released.

Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: You're right!
Would like a method equivalent with checking if the class exist?

Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: only use Type in this PR please. feel free to create a second PR with a solution for the future deprecation

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 10 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: @Tobion I don't understand what you are asking here. Can you clarify?

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 14]
```diff
@@ -108,7 +108,7 @@ public function send(string $body, array $headers, int $delay = 0): string
         $now = new \DateTime();
         $availableAt = (clone $now)->modify(sprintf('+%d seconds', $delay / 1000));
 
-        $queryBuilder = $this->driverConnection->createQueryBuilder()
+        $queryBuilder = $this->createQueryBuilder()
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: Why changing this? It's doing an insert, so the helper method seems useless here.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 13 above)
Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: It's still not consistent. Here you implemented the fallback https://github.com/symfony/symfony/blob/1e04345fd53b577141f57fe9df25ab9b6a249513/src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php#L346-L352 but https://github.com/symfony/symfony/blob/1e04345fd53b577141f57fe9df25ab9b6a249513/src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php#L325 does not use it. Either use the `getDateTimeType` helper everywhere, or don't add it at all.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 15]
```diff
@@ -122,8 +122,14 @@ public function send(string $body, array $headers, int $delay = 0): string
...
             $this->configuration['queue_name'],
-            self::formatDateTime($now),
-            self::formatDateTime($availableAt),
+            $now,
+            $availableAt,
+        ], [
+            null,
+            null,
+            null,
+            Type::getType($this->getDateTimeType()),
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: you don't need to call `Type::getType`. it should be enough to pass the string constant like it's done in other places https://github.com/symfony/symfony/blob/1e04345fd53b577141f57fe9df25ab9b6a249513/src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php#L285

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 10 above)
Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: https://github.com/symfony/symfony/pull/32456#discussion_r329619820

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 16]
```diff
@@ -329,15 +336,19 @@ private function getSchema(): Schema
...
         return $doctrineEnvelope;
     }
+
+    private function getDateTimeType()
+    {
+        if (!class_exists(\Doctrine\DBAL\Types\Types::class)) {
+            return Type::DATETIME;
+        }
+
+        return \Doctrine\DBAL\Types\Types::DATETIME_MUTABLE;
```
Comment (stof) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: please add a use statement

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 13 above)
Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: I asked to not use the new `Types` class because in order to fully fix the deprcation of `Type`, you would also need to add the fallback of the other types like "text" in https://github.com/symfony/symfony/blob/1e04345fd53b577141f57fe9df25ab9b6a249513/src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php#L321 which you did not. So it's not complete and irrelevant to this PR.

**Comment by roukmoute** [comment]
Ping @Tobion @stof

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 17]
```diff
@@ -280,12 +292,11 @@ private function createQueryBuilder(): QueryBuilder
             ->from($this->configuration['table_name'], 'm');
     }
 
-    private function executeQuery(string $sql, array $parameters = [])
+    private function executeQuery(string $sql, array $parameters = [], array $types = [])
     {
         $stmt = null;
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: no needed anymore

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 18]
```diff
@@ -124,8 +124,14 @@ public function send(string $body, array $headers, int $delay = 0): string
...
             $this->configuration['queue_name'],
-            self::formatDateTime($now),
-            self::formatDateTime($availableAt),
+            $now,
+            $availableAt,
+        ], [
+            null,
+            null,
+            null,
+            Type::getType(Type::DATETIME),
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: In https://github.com/symfony/symfony/pull/32456#discussion_r329623629 I asked to only pass the constant. No need to call `Type::getType` anywhere. Please it's very painful to review this PR over and over and the same problems keep being there.

**Review by Tobion — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 17 above)
Comment (roukmoute) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: Are you sure?
image

Comment (Tobion) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: I commented on the line `$stmt = null;`. The initialization is not needed anymore because it's always set.

**Review by Tobion — APPROVED** [review]

**Review by Tobion — CHANGES_REQUESTED** [review]
Please squash all commits to a single commit. Our tool cannot do it because of different commiters.

**Review by Tobion — APPROVED** [review]

**Comment by Tobion** [comment]
... and one more bug fixed, thank you @roukmoute.

**Comment by roukmoute** [comment]
Thanks for all your time @Tobion
